src/recipe/recipe.service.ts

```diff
@@ -35,35 +57,123 @@ export class RecipeService {
+      );
+    }
+
+    if (user.id != recipe.authorId && !recipe.isPublic) {
+      throw new ForbiddenException();
+    }
+
     return recipe;
   }
 
-  fetchAllRecipes(): Promise<Recipe[]> {
-    return this.prisma.recipe.findMany();
+  async fetchRecipesByAuthorId(
+    authorId: number,
+    principalId: number,
+  ): Promise<Recipe[]> {
+    const author = await this.prisma.user.findUnique({
+      where: { id: authorId },
+    });
+    if (!author) {
+      throw new NotFoundException();
+    }
+    const principal = await this.prisma.user.findUnique({
+      where: { id: principalId },
+    });
+
+    if (author.id === principal.id || principal.role == Role.ADMIN) {
+      return this.prisma.recipe.findMany({
+        where: {
+          authorId: author.id,
+        },
+      });
+    }
+
+    return this.prisma.recipe.findMany({
+      where: {
+        authorId: author.id,
+        isPublic: true,
+      },
+    });
+  }
+
+  async fetchAllRecipes(userId: number): Promise<Recipe[]> {
+    const user = await this.prisma.user.findUnique({
+      where: {
+        id: userId,
+      },
+    });
+    if (user.role === Role.ADMIN) {
+      return this.prisma.recipe.findMany();
+    } else {
+      return this.prisma.recipe.findMany({
+        where: {
+          isPublic: true,
+        },
+      });
+    }
   }
 
   async updateRecipe(
-    id: number,
+    userId: number,
+    recipeId: number,
     payload: UpdateRecipeRequest,
   ): Promise<Recipe> {
+    const user = await this.prisma.user.findUnique({
+      where: {
+        id: userId,
+      },
+    });
+    let recipe = await this.prisma.recipe.findUnique({
+      where: {
+        id: recipeId,
+      },
+    });
+
     try {
-      const recipe = await this.prisma.recipe.update({
-        where: { id },
+      if (user.id != recipe.authorId && user.role != Role.ADMIN) {
+        throw new ForbiddenException();
+      }
+
+      recipe = await this.prisma.recipe.update({
+        where: { id: recipeId },
         data: payload,
       });
       this.recipeCacheService.cacheRecipe(recipe);
-      return recipe;
     } catch {
       throw new NotFoundException();
     }
+
+    return recipe;
   }
 
-  async deleteRecipe(id: number): Promise<void> {
+  async deleteRecipe(recipeId: number, userId: number): Promise<void> {
+    const user = await this.prisma.user.findUnique({
+      where: {
+        id: userId,
+      },
+    });
+    const recipe = await this.prisma.recipe.findUnique({
+      where: {
+        id: recipeId,
+      },
+    });
     try {
+      if (user.id != recipe.authorId && user.role != Role.ADMIN) {
+        throw new ForbiddenException();
+      }
       await this.prisma.recipe.delete({
-        where: { id },
+        where: { id: recipeId },
       });
-      this.recipeCacheService.deleteCachedRecipe(id);
+      this.recipeCacheService.deleteCachedRecipe(recipeId);
     } catch {
       throw new NotFoundException();
     }
```

Comment:
I don't understand this part - you are throwing a `ForbiddenException` but it will be catched by `catch` block and `NotFoundException` will be thrown instead. Is there a need for a try catch here at all? I don't think Prisma returns error on missing record?

Reply:
`prisma.update` throws an `RecordNotFound` exception if record does not exist ([Prisma Client API reference](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference))